### PR TITLE
[SPARK-44066][SQL] Support positional parameters in Scala/Java `sql()`

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
In the PR, I propose to extend SparkSession API and override the `sql` method by:
```scala
  def sql(sqlText: String, args: Array[_]): DataFrame
```
which accepts an array of Java/Scala objects that can be converted to SQL literal expressions.

And the first argument `sqlText` might have named parameters in the positions of constants like literal values. A value can be also a `Column` of literal expression, in that case it is taken as is.

For example:
```scala
  spark.sql(
    sqlText = "SELECT * FROM tbl WHERE date > ? LIMIT ?",
    args = Array(LocalDate.of(2023, 6, 15), 100))
```
The new `sql()` method parses the input SQL statement and replaces the positional parameters by the literal values.

### Why are the changes needed?
1. To conform the SQL standard and JDBC/ODBC protocol.
2. To improve user experience with Spark SQL via
    - Using Spark as remote service (microservice).
    - Write SQL code that will power reports, dashboards, charts and other data presentation solutions that need to account for criteria modifiable by users through an interface.
    - Build a generic integration layer based on the SQL API. The goal is to expose managed data to a wide application ecosystem with a microservice architecture. It is only natural in such a setup to ask for modular and reusable SQL code, that can be executed repeatedly with different parameter values.

3. To achieve feature parity with other systems that support positional parameters.

### Does this PR introduce _any_ user-facing change?
No, the changes extend the existing API.

### How was this patch tested?
By running new tests:
```
$ build/sbt "test:testOnly *AnalysisSuite"
$ build/sbt "test:testOnly *PlanParserSuite"
$ build/sbt "test:testOnly *ParametersSuite"
```
and the affected test suites:
```
$ build/sbt "sql/testOnly *QueryExecutionErrorsSuite"
```

### CODE REVIEW

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala [hunk 1]
```diff
@@ -1360,22 +1360,40 @@ class AnalysisSuite extends AnalysisTest with Matchers {
...
 
+  test("SPARK-44066: bind positional parameters to literals") {
+    CTERelationDef.curId.set(0)
+    val actual1 = PosParameterizedQuery(
+      child = parsePlan("WITH a AS (SELECT 1 c) SELECT * FROM a LIMIT ?"),
+      args = Seq(Literal(10))).analyze
+    CTERelationDef.curId.set(0)
+    val expected1 = parsePlan("WITH a AS (SELECT 1 c) SELECT * FROM a LIMIT 10").analyze
+    comparePlans(actual1, expected1)
+    // Ignore unused arguments
```
Comment (entong) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:1387]: Can we add more negative test cases?
- Error out when there are less number of arguments than positional placeholders (?)
- Error out when the query text contains both named and positional params

Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:1387]: > Error out when there are less number of arguments than positional placeholders (?)

Such test has been added already, see `test("non-substituted positional parameters")` in `ParametersSuite`, please.

Comment (MaxGekk) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/analysis/AnalysisSuite.scala:1387]: > Error out when the query text contains both named and positional params

I added such test "mixing of positional and named parameters". Thank you for the proposal @entong. BTW, existing API doesn't allow to pass both named and positional arguments but the query might contain such parameters. The test checks that.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala [hunk 2]
```diff
@@ -71,21 +119,24 @@ object BindParameters extends Rule[LogicalPlan] with QueryErrorsBase {
...
-      case p @ ParameterizedQuery(child, args) if !child.containsPattern(UNRESOLVED_WITH) =>
-        args.find(!_._2.isInstanceOf[Literal]).foreach { case (name, expr) =>
-          expr.failAnalysis(
-            errorClass = "INVALID_SQL_ARG",
-            messageParameters = Map("name" -> name))
-        }
+      case p @ NameParameterizedQuery(child, args) if !child.containsPattern(UNRESOLVED_WITH) =>
+        checkArgs(args)
+        val res = bind(child) { case NamedParameter(name) if args.contains(name) => args(name) }
+        res.copyTagsFrom(p)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:125]: do we really need this? the function `bind` calls `resolveExpressionsWithPruning` which automatically propagate the tree node tags

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala [hunk 3]
```diff
@@ -41,17 +36,56 @@ case class Parameter(name: String) extends LeafExpression with Unevaluable {
...
+ * The expression represents a named parameter that should be replaced by a literal.
+ *
+ * @param name The identifier of the parameter without the marker.
+ */
+case class NamedParameter(name: String) extends Parameter
+
+/**
+ * The expression represents a positional parameter that should be replaced by a literal.
+ *
+ * @param pos An unique position of the parameter in a SQL query.
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:53]:
```suggestion
 * @param pos An unique position of the parameter in a SQL query text.
```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala [hunk 4]
```diff
@@ -71,21 +119,24 @@ object BindParameters extends Rule[LogicalPlan] with QueryErrorsBase {
...
+        res.copyTagsFrom(p)
+        res
+
+      case p @ PosParameterizedQuery(child, args) if !child.containsPattern(UNRESOLVED_WITH) =>
+        val indexedArgs = args.zipWithIndex
+        checkArgs(indexedArgs.map(arg => (s"_${arg._2}", arg._1)))
+
+        val positions = scala.collection.mutable.Set.empty[Int]
+        bind(child) { case p @ PosParameter(pos) => positions.add(pos); p }
+        val posToIndex = positions.toSeq.sorted.zipWithIndex.toMap
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:134]: shall we fail earlier if the number of parameters does not match the number of actual arguments?

Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:134]: not exact match, when the number of parameters is too less.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala (hunk 2 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:125]: it was probably an oversight...

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala (hunk 3 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:53]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala (hunk 2 above)
Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:125]: I removed it.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala (hunk 4 above)
Comment (cloud-fan) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/parameters.scala:132]: maybe it's fine to fail with unbound parameters.

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
Merging to master. Thank you, @entong @cloud-fan for review.

**Pull request closed without merging**